File: cpp/ml_methods/hac_clustering/demo_mlm_hac_02_mmap/demo_mlm_hac_02_mmap.cpp
// ==================================================================================
// demo_mlm_hac_02_mmap.cpp
// ==================================================================================
/**
 * @file demo_mlm_hac_02_mmap.cpp
 *
 * @brief LSH Linkage Comparison at biobank-representative scale via memory-mapped I/O.
 *
 * @details Replicates the geometry of demo_02_hac's `demo_lsh_linkage_comparison` at
 * extreme scale (N=100,000 samples, P=400,000 variables, K=10 clusters) without ever
 * allocating the full ~298 GB matrix in RAM. The dataset is generated and standardised
 * in-place via a memory-mapped file; the OS page cache streams data on demand.
 *
 * The demo is organised into three sequential phases:
 *
 *   Phase 1 | Generation   : Latent-factor correlated blocks are written column-by-column
 *                            into the mmap file in parallel.  RAM cost is bounded by the
 *                            latent signal matrix (N × K ≈ 8 MB) and the OS page cache.
 *
 *   Phase 2 | Standardise  : Each column is mean-centred and L2-normalised in-place
 *                            (required for Pearson-correlation distances).
 *
 *   Phase 3 | Clustering   : Linkage strategies are benchmarked under the pure
 *                            O(1) LSH Correlation Approximation:
 *                              A. Single Linkage   – SLINK (projected engine)
 *                              B. Complete Linkage – [SKIPPED] O(p^2) Lance-Williams
 *                                 matrix is intractable at this scale
 *                              C. Average Linkage  – UPGMA (projected engine)
 *                              D. Ward Linkage      – minimum-variance (projected engine)
 *                            WPGMA is likewise skipped as intractable at p = 400,000.
 *
 * Expected result: the executed methods achieve ARI ≈ 1.0000 because the block geometry
 * creates well-separated clusters in correlation space; the O(1) SimHash approximation
 * recovers the exact solution without paying the O(P) exact-distance cost.
 *
 * @note  The memory-mapped file path is injected at compile time via the CMake
 *        TREX_DEMO_MMAP_PATH definition (defaults to a relative path if not set).
 *        The file is removed automatically on clean exit.
 */
// ==================================================================================

// std includes
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <random>
#include <string>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// ml_methods includes
#include <ml_methods/clustering/hierarchical/agglomerative/agglomerative_dispatcher.hpp>
#include <ml_methods/clustering/hierarchical/agglomerative/dendrogram_utils.hpp>
#include <ml_methods/clustering/hierarchical/agglomerative/distance_policy.hpp>
#include <ml_methods/clustering/hierarchical/agglomerative/generic_linkage_core.hpp>
#include <ml_methods/clustering/hierarchical/agglomerative/nnchain_core.hpp>
#include <ml_methods/clustering/hierarchical/agglomerative/slink_core.hpp>

// utils includes
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>
#include <utils/memmap/memory_mapped_matrix.hpp>
#include <utils/openmp/utils_openmp.hpp>

// ==================================================================================
// Namespace aliases
// ==================================================================================

namespace hac      = trex::ml_methods::clustering::hierarchical::agglomerative;
namespace memmap   = trex::utils::memmap;
namespace omp_utils = trex::utils::openmp;
namespace cdiagnostics = trex::utils::eval::cdiagnostics;

// Type aliases for the memory-mapped matrix and its Eigen views
using MmapMatrix    = memmap::MemoryMappedMatrix<double>;
using EigenMap      = MmapMatrix::MapType;        ///< Eigen::Map<MatrixXd, ColMajor> (mutable)
using ConstEigenMap = MmapMatrix::ConstMapType;   ///< Eigen::Map<const MatrixXd, ColMajor>


// ==================================================================================
// Part 1: Data Generation, Preprocessing, and Evaluation
// ==================================================================================

// ----------------------------------------------------------------------------------

/**
 * @brief Computes the Adjusted Rand Index (ARI) to evaluate clustering accuracy.
 *
 * @param true_labels Ground-truth cluster assignments (size P).
 * @param pred_labels Predicted cluster assignments (size P).
 * @return ARI in [-1, 1]; perfect agreement yields 1.0.
 */
double compute_ari(
    const std::vector<Eigen::Index>& true_labels,
    const std::vector<Eigen::Index>& pred_labels)
{
    if (true_labels.size() != pred_labels.size() || true_labels.empty()) return 0.0;

    const Eigen::Index n        = static_cast<Eigen::Index>(true_labels.size());
    const Eigen::Index max_true = *std::max_element(true_labels.begin(),
                                                    true_labels.end());
    const Eigen::Index max_pred = *std::max_element(pred_labels.begin(),
                                                    pred_labels.end());

    // Contingency table
    Eigen::MatrixXd contingency = Eigen::MatrixXd::Zero(max_true + 1, max_pred + 1);
    for (Eigen::Index i = 0; i < n; ++i)
        contingency(true_labels[i], pred_labels[i])++;

    auto choose2 = [](double x) { return x * (x - 1.0) / 2.0; };

    double sum_comb_c = 0.0;
    for (Eigen::Index i = 0; i <= max_true; ++i)
        for (Eigen::Index j = 0; j <= max_pred; ++j)
            sum_comb_c += choose2(contingency(i, j));

    double sum_comb_a = 0.0;
    for (Eigen::Index i = 0; i <= max_true; ++i)
        sum_comb_a += choose2(contingency.row(i).sum());

    double sum_comb_b = 0.0;
    for (Eigen::Index i = 0; i <= max_pred; ++i)
        sum_comb_b += choose2(contingency.col(i).sum());

    const double n_choose_2     = choose2(static_cast<double>(n));
    const double expected_index = (sum_comb_a * sum_comb_b) / n_choose_2;
    const double max_index      = (sum_comb_a + sum_comb_b) / 2.0;

    if (max_index == expected_index) return 1.0;
    return (sum_comb_c - expected_index) / (max_index - expected_index);
}

// ----------------------------------------------------------------------------------

/** @brief Configuration parameters for the synthetic data generator. */
struct DataGenConfig {
    Eigen::Index n_samples;
    Eigen::Index p_features;
    Eigen::Index k_clusters;
};

// ----------------------------------------------------------------------------------

/**
 * @brief Computes per-feature ground-truth cluster labels from block sizes.
 *
 * @param block_sizes Exact number of features per cluster (size = k_clusters).
 * @param k_clusters  Number of clusters.
 * @return Vector of length P containing the block label for each feature.
 */
std::vector<Eigen::Index> build_true_labels(
    const std::vector<Eigen::Index>& block_sizes,
    Eigen::Index                      k_clusters)
{
    Eigen::Index total_features = 0;
    for (auto s : block_sizes) total_features += s;

    std::vector<Eigen::Index> true_labels(total_features);
    Eigen::Index current = 0;
    for (Eigen::Index k = 0; k < k_clusters; ++k)
        for (Eigen::Index j = 0; j < block_sizes[k]; ++j)
            true_labels[current++] = k;

    return true_labels;
}

// ----------------------------------------------------------------------------------

/**
 * @brief Generates correlated variable blocks directly into a memory-mapped matrix.
 *
 * @details Uses a Latent Factor Model:
 * @code
 *   X(i, j) = mu_k + sigma_k * [ signal * latent(i, k) + noise * epsilon(i, j) ]
 * @endcode
 * where `k = block(j)`. The latent signal matrix (N × K ≈ 8 MB) is held in RAM;
 * all N × P features are written column-by-column into the mmap, never requiring
 * the full matrix in RAM simultaneously.
 *
 * @param map         Writable Eigen::Map onto the memory-mapped file (N × P, ColMajor).
 * @param config      Structural dimensions {n_samples, p_features, k_clusters}.
 * @param block_sizes Exact number of features per cluster (must sum to p_features).
 * @param true_labels Pre-computed per-feature block labels (size = p_features).
 * @param centers     Per-block mean offsets (size = k_clusters).
 * @param std_devs    Per-block scale multipliers (size = k_clusters).
 */
void generate_correlated_variables_to_mmap(
    EigenMap&                          map,
    const DataGenConfig&               config,
    const std::vector<Eigen::Index>&   true_labels,
    const std::vector<double>&         centers,
    const std::vector<double>&         std_devs)
{
    // -- Latent signals: (N × K) in RAM  (~8 MB for N=100k, K=10) -----------------
    Eigen::MatrixXd latent(config.n_samples, config.k_clusters);
    {
        std::mt19937 rng(42);
        std::normal_distribution<double> dist(0.0, 1.0);
        for (Eigen::Index b = 0; b < config.k_clusters; ++b)
            for (Eigen::Index i = 0; i < config.n_samples; ++i)
                latent(i, b) = dist(rng);
    }

    constexpr double signal_strength = 0.85;
    constexpr double noise_strength  = 0.15;

    // -- Stream features column-by-column into the mmap (ColMajor: cache-optimal) -
    #pragma omp parallel
    {
        const int thread_id = omp_get_thread_num();
        std::mt19937 local_rng(495 + thread_id);
        std::normal_distribution<double> local_dist(0.0, 1.0);

        #pragma omp for schedule(static)
        for (Eigen::Index j = 0; j < config.p_features; ++j) {
            const Eigen::Index blk  = true_labels[j];
            const double       mu   = centers[blk];
            const double       sigma = std_devs[blk];

            for (Eigen::Index i = 0; i < config.n_samples; ++i) {
                double base = signal_strength * latent(i, blk)
                            + noise_strength  * local_dist(local_rng);
                map(i, j) = mu + sigma * base;
            }
        }
    }
}

// ----------------------------------------------------------------------------------

/**
 * @brief In-place mean-centring and L2-normalisation (Pearson-correlation scaling).
 *
 * @details Operates column-by-column which is cache-friendly for ColMajor matrices
 * and sequential-streaming-optimal for memory-mapped files.
 *
 * @param mat Any Eigen dense expression with writable columns.
 */
template <typename Derived>
void standardize_for_correlation(Eigen::MatrixBase<Derived>& mat) {
    #pragma omp parallel for schedule(static)
    for (Eigen::Index j = 0; j < mat.cols(); ++j) {
        const double mean = mat.col(j).mean();
        mat.col(j).array() -= mean;

        const double norm = mat.col(j).norm();
        if (norm > 1e-12)
            mat.col(j) /= norm;
        else
            mat.col(j).setZero();
    }
}


// ==================================================================================
// Part 2: Demo
// ==================================================================================

/**
 * @brief Mmap-backed LSH Linkage Comparison at biobank-representative scale.
 *
 * Generates N=100,000 × P=400,000 correlated variables in K=10 imbalanced blocks,
 * standardises them, then benchmarks four HAC linkage strategies under the O(1)
 * LSH Correlation Approximation. Disk footprint ≈ 320 GB; RAM footprint is
 * determined by the OS page cache and the O(P) HAC bookkeeping structures.
 */
void demo_lsh_linkage_comparison_mmap() {
    cdiagnostics::print_section_header(
        "Demo 03: LSH Linkage Comparison via Memory-Mapped I/O (N=100k, P=400k, K=10)");

    // -------------------------------------------------------------------------
    // 0. Configuration
    // -------------------------------------------------------------------------

    // File path injected by CMake; falls back to relative path if not set.
#ifdef TREX_DEMO_MMAP_PATH
    const std::string MMAP_PATH = TREX_DEMO_MMAP_PATH;
#else
    const std::string MMAP_PATH = "trex_demo03_hac_mmap.bin";
#endif

    const DataGenConfig config {
        .n_samples  = 100'000,
        .p_features = 400'000,
        .k_clusters = 10
    };

    // Imbalanced block sizes — 10 blocks summing exactly to 400,000.
    // The deliberate imbalance (ranging from 33k to 50k features) tests
    // robustness of the LSH approximation to non-uniform cluster sizes.
    const std::vector<Eigen::Index> block_sizes = {
        50'000, 48'000, 45'000, 42'000, 40'000,
        38'000, 35'000, 33'000, 35'000, 34'000
    };
    static_assert(
        50'000 + 48'000 + 45'000 + 42'000 + 40'000 +
        38'000 + 35'000 + 33'000 + 35'000 + 34'000 == 400'000,
        "block_sizes must sum to p_features"
    );

    // Per-block cluster centres — spread across [-5.0, +5.0]
    const std::vector<double> centers = {
         0.0,  2.0, -2.0,  4.0, -4.0,
         1.5, -1.5,  3.0, -3.0,  5.0
    };

    // Per-block noise scales — varied to create realistic heteroscedasticity
    const std::vector<double> std_devs = {
        1.5, 1.2, 0.8, 2.0, 3.5,
        1.0, 1.8, 0.6, 2.5, 1.3
    };

    // Precompute true labels — deterministic from block_sizes, O(P)
    const std::vector<Eigen::Index> true_labels =
        build_true_labels(block_sizes, config.k_clusters);

    // Disk footprint
    const std::size_t file_bytes =
        static_cast<std::size_t>(config.n_samples) *
        static_cast<std::size_t>(config.p_features) * sizeof(double);

    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Memory-mapped file : " << MMAP_PATH << "\n";
    std::cout << "File size          : "
              << static_cast<double>(file_bytes) / static_cast<double>(1ULL << 30)
              << " GB\n";
    std::cout << "Latent RAM (N x K)   : "
              << static_cast<double>(config.n_samples * config.k_clusters * sizeof(double))
                  / (1ULL << 20)
              << " MB\n\n";


    // -------------------------------------------------------------------------
    // Phase 1 — Data generation (ReadWrite)
    // Phase 2 — Standardisation (in-place, same ReadWrite mapping)
    // Skipped when the mmap file already exists and has the expected size.
    // Note: MmapMatrix pre-allocates the file to its full size up front, so this
    // size check cannot distinguish a complete file from one left behind by a run
    // that crashed mid-Phase-1; a stale partial file would be silently reused.
    // -------------------------------------------------------------------------
    if (!std::filesystem::exists(MMAP_PATH) ||
         std::filesystem::file_size(MMAP_PATH) != file_bytes) {
        std::cout << "=== Phase 1: Generating correlated variables into mmap ===\n";
        std::cout << "    N=" << config.n_samples
                  << ", P=" << config.p_features
                  << ", K=" << config.k_clusters << "\n";

        MmapMatrix mmap(MMAP_PATH,
                        static_cast<std::size_t>(config.n_samples),
                        static_cast<std::size_t>(config.p_features),
                        memmap::AccessMode::ReadWrite);
        EigenMap X_rw = mmap.getMap();

        const double t0 = omp_get_wtime();
        generate_correlated_variables_to_mmap(
            X_rw, config, true_labels, centers, std_devs);
        const double t1 = omp_get_wtime();

        std::cout << "    Done. (" << std::setprecision(1) << (t1 - t0) << " s)\n\n";

        std::cout << "=== Phase 2: Standardising columns (mean=0, L2=1) in-place ===\n";

        const double t2 = omp_get_wtime();
        standardize_for_correlation(X_rw);
        const double t3 = omp_get_wtime();

        std::cout << "    Done. (" << std::setprecision(1) << (t3 - t2) << " s)\n\n";

        // MmapMatrix destructor closes and flushes the mapping here (RAII).
    } else {
        std::cout << "=== Phases 1 & 2: Skipped (mmap file exists with correct size) ===\n\n";
    }


    // -------------------------------------------------------------------------
    // Phase 3 — Clustering (reopen ReadWrite; HAC does not mutate the data)
    // -------------------------------------------------------------------------
    std::cout << "=== Phase 3: Reopening mmap for clustering ===\n";

    MmapMatrix mmap_cluster(MMAP_PATH,
                            static_cast<std::size_t>(config.n_samples),
                            static_cast<std::size_t>(config.p_features),
                            memmap::AccessMode::ReadWrite);
    EigenMap X_map = mmap_cluster.getMap();

    // Pure O(1) LSH Correlation Approximation (64-bit SimHash)
    std::cout << "    Distance policy: Correlation LSH Approximation (O(1) SimHash)\n\n";

    using LshDistPol = hac::DistancePolicy<
        EigenMap, hac::DistanceMetric::Correlation_LSH_Approx
    >;

    // Test A — Single Linkage via Dispatcher (Routes to SLINK)
    // -------------------------------------------------------------------------
    {
        std::cout << "[Test A] Single Linkage (SLINK + LSH):\n";

        const double t0 = omp_get_wtime();
        auto merges = hac::AgglomerativeClustering::cluster<
            EigenMap, LshDistPol, hac::LinkageMethod::Single>(X_map);
        const double t1 = omp_get_wtime();

        auto labels = hac::DendrogramUtils::cut_tree(
            merges, X_map.cols(), config.k_clusters);
        const double ari = compute_ari(true_labels, labels);

        std::cout << "    Time : " << std::setprecision(4) << (t1 - t0) << " s\n";
        std::cout << "    ARI  : " << std::setprecision(4) << ari << " / 1.0000\n\n";
    }

    // Test B — Complete Linkage & WPGMA (Mathematically Banned at this scale)
    // -------------------------------------------------------------------------
    {
        std::cout << "[Test B] Complete Linkage & WPGMA:\n";
        std::cout << "    [SKIPPED] These methods lack a geometric center of mass.\n";
        std::cout << "    They strictly require the O(P^2) Lance-Williams matrix.\n";
        std::cout << "    At P=400,000, this requires 1.2 TB of NVMe I/O, which is intractable.\n\n";
    }

    // Test C — Average Linkage (UPGMA) via Dispatcher
    // Routes to Projected 64D RAM Engine!
    // -------------------------------------------------------------------------
    {
        std::cout << "[Test C] UPGMA Average Linkage (Projected 64D RAM Engine + LSH):\n";

        const double t0 = omp_get_wtime();
        // Notice we don't need use_mmap=true anymore. The dispatcher intercepts
        // this and builds the compressed 204 MB RAM matrix.
        auto merges = hac::AgglomerativeClustering::cluster<
            EigenMap, LshDistPol, hac::LinkageMethod::Average>(X_map);
        const double t1 = omp_get_wtime();

        auto labels = hac::DendrogramUtils::cut_tree(
            merges, X_map.cols(), config.k_clusters);
        const double ari = compute_ari(true_labels, labels);

        std::cout << "    Time : " << std::setprecision(4) << (t1 - t0) << " s\n";
        std::cout << "    ARI  : " << std::setprecision(4) << ari << " / 1.0000\n\n";
    }

    // Test D — Ward's Minimum Variance via Dispatcher
    // Routes to Projected 64D RAM Engine!
    // -------------------------------------------------------------------------
    {
        std::cout << "[Test D] Ward's Minimum Variance (Projected 64D RAM Engine + LSH):\n";

        const double t0 = omp_get_wtime();
        // The dispatcher intercepts this as well and runs it purely in RAM.
        auto merges = hac::AgglomerativeClustering::cluster<
            EigenMap, LshDistPol, hac::LinkageMethod::Ward>(X_map);
        const double t1 = omp_get_wtime();

        auto labels = hac::DendrogramUtils::cut_tree(
            merges, X_map.cols(), config.k_clusters);
        const double ari = compute_ari(true_labels, labels);

        std::cout << "    Time : " << std::setprecision(4) << (t1 - t0) << " s\n";
        std::cout << "    ARI  : " << std::setprecision(4) << ari << " / 1.0000\n\n";
    }

    std::cout << "[Demo 03 Complete]\n\n";


    // -------------------------------------------------------------------------
    // Cleanup: remove the mmap file
    // -------------------------------------------------------------------------
    {
        namespace fs = std::filesystem;
        std::error_code ec;
        if (fs::remove(MMAP_PATH, ec))
            std::cout << "Cleaned up mmap file: " << MMAP_PATH << "\n";
        else
            std::cerr << "[WARN] Could not remove mmap file: " << ec.message() << "\n";
    }
}



// ==================================================================================
// 3. Main
// ==================================================================================

void init_omp_environment(int threads = 6) {
    omp_set_num_threads(threads);
    omp_set_schedule(omp_sched_static, 0);
}

int main() {
    std::cout << std::unitbuf; // Flush after every write — keeps log file current.
    std::cout << "\n";
    cdiagnostics::print_section_header(
        "HAC Memory-Mapped Demo Suite (N=100k x P=400k)");

    omp_utils::print_info();
    init_omp_environment(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    try {
        demo_lsh_linkage_comparison_mmap();
        cdiagnostics::print_section_header("Demo 03 completed successfully");
    } catch (const std::exception& e) {
        std::cerr << "\n[ERROR] " << e.what() << "\n\n";
        return 1;
    }

    return 0;
}

File: Python/ml_methods/hac_clustering/demo_mlm_hac_02_mmap/demo_mlm_hac_02_mmap.py
# ==============================================================================
# demo_mlm_hac_02_mmap.py
# ==============================================================================
#
# LSH Linkage Comparison on memory-mapped data.
#
# Mirrors cpp/ml_methods/hac_clustering/demo_mlm_hac_02_mmap/
#         demo_mlm_hac_02_mmap.cpp.
#
# The cpp demo replicates the LSH linkage comparison at biobank scale
# (N = 100,000 samples, P = 400,000 variables, K = 10 clusters, ~298 GB on
# disk) without ever allocating the full matrix in RAM. This Python mirror
# keeps the same three-phase out-of-core workflow at an illustrative scale:
# N = 2,000, P = 4,000, K = 10 (~61 MB on disk).
#
# The demo is organized into three sequential phases:
#
#   Phase 1 | Generation   : Latent-factor correlated blocks are written
#                            variable-by-variable into the mmap file. RAM cost
#                            is bounded by the latent signal matrix (N x K)
#                            plus one variable at a time.
#
#   Phase 2 | Standardize  : Each variable is mean-centered and L2-normalized
#                            in place (required for Pearson-correlation
#                            distances).
#
#   Phase 3 | Clustering   : Linkage strategies are benchmarked under the pure
#                            O(1) LSH Correlation Approximation:
#                              A. Single Linkage   - SLINK
#                              B. Complete/WPGMA   - [SKIPPED] the O(p^2)
#                                 Lance-Williams matrix is intractable at the
#                                 cpp scale (P = 400,000); the skip is kept to
#                                 mirror the cpp console output.
#                              C. Average Linkage  - UPGMA (projected engine)
#                              D. Ward Linkage     - minimum variance
#                                                    (projected engine)
#
# Expected result: the executed methods achieve ARI ~ 1.0000 because the block
# geometry creates well-separated clusters in correlation space.
#
# Notes on the port:
#   - agglomerative_cluster() clusters the ROWS of its input, so the matrix is
#     stored TRANSPOSED on disk relative to cpp: P x N, one variable per row.
#     The zero-copy to_numpy() view of the mmap file is passed directly to the
#     clustering call; use_mmap=True additionally keeps the backend's
#     intermediate distance data on disk.
#   - The cpp demo generates and standardizes with OpenMP parallel loops
#     (C++-exclusive) and skips Phases 1-2 when a file of the expected size
#     already exists; this mirror uses a fresh temporary file every run.
#
# ==============================================================================

import os
import tempfile
import time

import numpy as np

from trex_selector_neo.ml_methods.clustering import (
    DistanceMetric,
    LinkageMethod,
    agglomerative_cluster,
    cut_tree,
)
from trex_selector_neo.utils import (
    AccessMode,
    MemoryMappedMatrix,
    get_max_threads,
    set_num_threads,
)


def print_section_header(title):
    print("=" * 78)
    print(f" {title}")
    print("=" * 78 + "\n")


# ==============================================================================
# Part 1: Data Generation, Preprocessing, and Evaluation
# ==============================================================================

def compute_ari(true_labels, pred_labels):
    """Adjusted Rand Index (ARI) in [-1, 1]; perfect agreement yields 1.0."""
    true_labels = np.asarray(true_labels)
    pred_labels = np.asarray(pred_labels)
    if true_labels.size == 0 or true_labels.size != pred_labels.size:
        return 0.0

    n = true_labels.size
    contingency = np.zeros((true_labels.max() + 1, pred_labels.max() + 1))
    np.add.at(contingency, (true_labels, pred_labels), 1.0)

    def choose2(x):
        return x * (x - 1.0) / 2.0

    sum_comb_c = choose2(contingency).sum()
    sum_comb_a = choose2(contingency.sum(axis=1)).sum()
    sum_comb_b = choose2(contingency.sum(axis=0)).sum()

    expected_index = (sum_comb_a * sum_comb_b) / choose2(float(n))
    max_index = (sum_comb_a + sum_comb_b) / 2.0

    if max_index == expected_index:
        return 1.0
    return (sum_comb_c - expected_index) / (max_index - expected_index)


def build_true_labels(block_sizes):
    """Per-variable ground-truth cluster labels from the block sizes."""
    return np.repeat(np.arange(len(block_sizes)), block_sizes)


def generate_correlated_variables_to_mmap(view, n_samples, k_clusters,
                                          true_labels, centers, std_devs,
                                          seed=42):
    """Generate correlated variable blocks directly into the mmap view.

    Latent factor model: X(j, i) = mu_k + sigma_k * [signal * latent(i, k) +
    noise * eps(j, i)] with k = block(j). The view is P x N (one variable per
    row); only the latent matrix (N x K) and one variable at a time are held
    in RAM.
    """
    rng = np.random.default_rng(seed)
    latent = rng.standard_normal((n_samples, k_clusters))  # N x K in RAM

    signal_strength = 0.85
    noise_strength = 0.15

    p_features = view.shape[0]
    for j in range(p_features):
        blk = true_labels[j]
        base = (signal_strength * latent[:, blk]
                + noise_strength * rng.standard_normal(n_samples))
        view[j, :] = centers[blk] + std_devs[blk] * base


def standardize_rows_for_correlation(view):
    """In-place mean-centering and L2-normalization, one variable (row) at a
    time — sequential-streaming-friendly for the memory-mapped file."""
    for j in range(view.shape[0]):
        row = view[j, :]
        row -= row.mean()
        norm = np.linalg.norm(row)
        if norm > 1e-12:
            row /= norm
        else:
            row[:] = 0.0
        view[j, :] = row


# ==============================================================================
# Part 2: Demo
# ==============================================================================

def demo_lsh_linkage_comparison_mmap():
    """Mmap-backed LSH Linkage Comparison.

    Generates P x N correlated variables in K imbalanced blocks directly in a
    memory-mapped file, standardizes them in place, then benchmarks HAC
    linkage strategies under the O(1) LSH Correlation Approximation.
    """
    print_section_header(
        "Demo 03: LSH Linkage Comparison via Memory-Mapped I/O "
        "(N=2k, P=4k, K=10)")

    # -------------------------------------------------------------------------
    # 0. Configuration
    # -------------------------------------------------------------------------

    # cpp: N = 100,000, P = 400,000 (~298 GB); scaled by 1/50 and 1/100 to
    # N = 2,000, P = 4,000 (~61 MB) here.
    n_samples, p_features, k_clusters = 2_000, 4_000, 10

    # Imbalanced block sizes — 10 blocks summing exactly to 4,000 (cpp block
    # sizes scaled by 1/100). The deliberate imbalance tests robustness of the
    # LSH approximation to non-uniform cluster sizes.
    block_sizes = [500, 480, 450, 420, 400, 380, 350, 330, 350, 340]
    assert sum(block_sizes) == p_features, "block_sizes must sum to p_features"

    # Per-block cluster centers — spread across [-5.0, +5.0]
    centers = [0.0, 2.0, -2.0, 4.0, -4.0, 1.5, -1.5, 3.0, -3.0, 5.0]

    # Per-block noise scales — varied to create realistic heteroscedasticity
    std_devs = [1.5, 1.2, 0.8, 2.0, 3.5, 1.0, 1.8, 0.6, 2.5, 1.3]

    # Precompute true labels — deterministic from block_sizes, O(P)
    true_labels = build_true_labels(block_sizes)

    # Disk footprint (cpp injects the path via TREX_DEMO_MMAP_PATH; a
    # temporary file in the system temp directory is used here)
    fd, mmap_path = tempfile.mkstemp(prefix="trex_demo03_hac_mmap_",
                                     suffix=".bin")
    os.close(fd)
    file_bytes = n_samples * p_features * 8

    print(f"Memory-mapped file : {mmap_path}")
    print(f"File size          : {file_bytes / (1 << 20):.2f} MB")
    print(f"Latent RAM (N x K) : "
          f"{n_samples * k_clusters * 8 / (1 << 20):.2f} MB\n")

    try:
        # ---------------------------------------------------------------------
        # Phase 1 — Data generation (ReadWrite)
        # Phase 2 — Standardization (in-place, same ReadWrite mapping)
        # ---------------------------------------------------------------------
        print("=== Phase 1: Generating correlated variables into mmap ===")
        print(f"    N={n_samples}, P={p_features}, K={k_clusters}")

        # Stored transposed relative to cpp: P x N, one variable per row.
        mmap_matrix = MemoryMappedMatrix(mmap_path, p_features, n_samples,
                                         AccessMode.ReadWrite)
        X_map = mmap_matrix.to_numpy()  # zero-copy view into the file

        t0 = time.perf_counter()
        generate_correlated_variables_to_mmap(
            X_map, n_samples, k_clusters, true_labels, centers, std_devs,
            seed=42)
        print(f"    Done. ({time.perf_counter() - t0:.1f} s)\n")

        print("=== Phase 2: Standardizing variables (mean=0, L2=1) in-place ===")
        t0 = time.perf_counter()
        standardize_rows_for_correlation(X_map)
        print(f"    Done. ({time.perf_counter() - t0:.1f} s)\n")

        # ---------------------------------------------------------------------
        # Phase 3 — Clustering (HAC does not mutate the data)
        # ---------------------------------------------------------------------
        print("=== Phase 3: Clustering the memory-mapped variables ===")
        print("    Distance policy: Correlation LSH Approximation "
              "(O(1) SimHash)\n")

        def run_test(title, method):
            print(f"{title}:")
            t0 = time.perf_counter()
            merges = agglomerative_cluster(
                X_map, method=method,
                metric=DistanceMetric.Correlation_LSH_Approx,
                use_mmap=True)  # backend intermediates also stay on disk
            elapsed = time.perf_counter() - t0
            labels = cut_tree(merges, p_features, k_clusters)
            ari = compute_ari(true_labels, labels)
            print(f"    Time : {elapsed:.4f} s")
            print(f"    ARI  : {ari:.4f} / 1.0000\n")

        # Test A — Single Linkage (routes to SLINK)
        run_test("[Test A] Single Linkage (SLINK + LSH)", LinkageMethod.Single)

        # Test B — Complete Linkage & WPGMA (skipped, mirroring cpp)
        print("[Test B] Complete Linkage & WPGMA:")
        print("    [SKIPPED] These methods lack a geometric center of mass.")
        print("    They strictly require the O(P^2) Lance-Williams matrix.")
        print("    At the cpp scale (P=400,000) this requires 1.2 TB of NVMe "
              "I/O, which is intractable.\n")

        # Test C — Average Linkage (UPGMA, projected RAM engine)
        run_test("[Test C] UPGMA Average Linkage "
                 "(Projected 64D RAM Engine + LSH)", LinkageMethod.Average)

        # Test D — Ward's Minimum Variance (projected RAM engine)
        run_test("[Test D] Ward's Minimum Variance "
                 "(Projected 64D RAM Engine + LSH)", LinkageMethod.Ward)

        print("[Demo 03 Complete]\n")

        del X_map, mmap_matrix  # unmap before the file is removed
    finally:
        # ---------------------------------------------------------------------
        # Cleanup: remove the mmap file
        # ---------------------------------------------------------------------
        try:
            os.unlink(mmap_path)
            print(f"Cleaned up mmap file: {mmap_path}")
        except OSError as e:
            print(f"[WARN] Could not remove mmap file: {e}")


# ==============================================================================
# 3. Main
# ==============================================================================

if __name__ == "__main__":
    print()
    print_section_header("HAC Memory-Mapped Demo Suite (N=2k x P=4k)")

    # cpp initializes OpenMP with 6 threads; mirrored via the utils entry points.
    set_num_threads(min(6, get_max_threads()))
    print(f"Running with {get_max_threads()} threads\n")

    demo_lsh_linkage_comparison_mmap()

    print_section_header("Demo 03 completed successfully")
